File: services/extract/router.py
"""Phase 2: Field Router — match schema fields to the right chunks using schema hints.

For short documents (chunk count ≤ max_chunks_per_field), the heuristic
scorer in `route_fields` is sufficient — every chunk is within reach.

For long documents, the LLM section map (`build_section_map`) provides a
smarter first pass: it asks the LLM which chunks are relevant to which
fields, then `route_fields` uses that mapping instead of heuristics.
This is triggered automatically in the pipeline when the document is
large enough that heuristic routing would be forced to drop content.
"""

from __future__ import annotations

import json
import re
from dataclasses import dataclass, field

from .document_map import Chunk


# Type alias — the map pass returns {field_name: [chunk_index, ...]}
SectionMap = dict[str, list[int]]


@dataclass
class FieldRoute:
    """A field mapped to specific chunks for extraction."""

    field_name: str
    field_spec: dict
    chunks: list[Chunk]
    source: str  # "hint", "signal_inferred", "broadened", "fallback", "section_map"


# ── Generic signal inference ────────────────────────────────────────
# When the schema has no hints, infer what to look for based on field type.
# This is generic, not domain-specific.

TYPE_SIGNAL_MAP = {
    "date": ["has_dates"],
    "number": ["has_dollar_amounts", "has_key_value_pairs"],
    "string": ["has_key_value_pairs"],
    "enum": ["has_key_value_pairs"],
    "array": ["has_tables"],
}


def _score_chunk(chunk: Chunk, field_name: str, field_spec: dict, total_chunks: int = 1) -> float:
    """Score how likely a chunk is to contain a field. Uses schema hints or generic inference."""
    score = 0.0
    hints = field_spec.get("hints", {})

    # ── Hint-based scoring (user-defined, highest priority) ──

    look_in = hints.get("look_in", [])
    if look_in:
        if chunk.category in look_in:
            score += 15.0

    prefer_contains = hints.get("prefer_contains") or []
    if prefer_contains:
        haystack = f"{chunk.title} {chunk.content}".lower()
        for phrase in prefer_contains:
            if isinstance(phrase, str) and phrase and phrase.lower() in haystack:
                score += 15.0
                break

    prefer_position = hints.get("prefer_position")
    if prefer_position:
        # Positional bias: 0-10 point bonus that favors chunks near
        # the requested end of the document. "top" gives max bonus to
        # chunk 0 (the header), decaying linearly; "bottom" is the
        # mirror. Designed for fields like merchant_name (always at
        # the top) or signature blocks (always at the bottom).
        if total_chunks <= 1:
            frac = 0.0
        else:
            frac = chunk.index / (total_chunks - 1)
        if prefer_position == "top":
            score += 10.0 * (1.0 - frac)
        elif prefer_position == "bottom":
            score += 10.0 * frac

    patterns = hints.get("patterns", [])
    if patterns:
        text = f"{chunk.title} {chunk.content[:1500]}".lower()
        for pattern in patterns:
            if re.search(pattern, text, re.IGNORECASE):
                score += 8.0
                break

    signals = hints.get("signals", [])
    if signals:
        for signal in signals:
            if signal in chunk.signals:
                score += 4.0

    # If we had hints and scored, return early — hints are authoritative
    if hints and score > 0:
        return score

    # ── Generic inference (no hints provided) ──

    # Infer useful signals from field type
    field_type = field_spec.get("type", "string")
    inferred_signals = TYPE_SIGNAL_MAP.get(field_type, [])
    for signal in inferred_signals:
        if signal in chunk.signals:
            score += 2.0

    # Field name appears in chunk title or content (fuzzy)
    field_words = field_name.replace("_", " ").lower()
    if field_words in chunk.title.lower():
        score += 6.0
    elif field_words in chunk.content[:500].lower():
        score += 3.0

    # Individual words from field name
    words = field_name.lower().split("_")
    text = f"{chunk.title} {chunk.content[:500]}".lower()
    word_hits = sum(1 for w in words if w in text and len(w) > 2)
    score += word_hits * 1.5

    return score


def _field_max_chunks(field_spec: dict, default: int) -> int:
    """Get the max chunks to route for this field.

    Per-field override via `hints.max_chunks` lets schema authors tell
    the router that a field legitimately needs to aggregate data from
    many chunks. Array-of-objects fields (like a list of policies
    spread across detail sections) often need this.
    """
    hints = field_spec.get("hints") or {}
    override = hints.get("max_chunks")
    if isinstance(override, int) and override > 0:
        return override
    return default


def route_fields(
    schema_def: dict,
    chunks: list[Chunk],
    max_chunks_per_field: int = 3,
    section_map: SectionMap | None = None,
) -> list[FieldRoute]:
    """Route each schema field to the most relevant chunks.

    `max_chunks_per_field` is the default cap. Individual fields can
    override via `hints.max_chunks` in the schema — useful for array
    fields that aggregate data across many detail sections.

    When `section_map` is provided (from `build_section_map`), it takes
    priority: fields use the LLM-assigned chunks directly, with no cap.
    Fields not in the map or mapped to empty lists fall through to the
    heuristic scorer so nothing is silently dropped.
    """
    fields = schema_def.get("fields", {})
    routes: list[FieldRoute] = []
    chunk_by_index = {c.index: c for c in chunks}

    for field_name, field_spec in fields.items():
        has_hints = bool(field_spec.get("hints"))
        field_cap = _field_max_chunks(field_spec, max_chunks_per_field)

        # ── Section map path (LLM-assigned chunks) ──
        if section_map and field_name in section_map and section_map[field_name]:
            mapped_chunks = [
                chunk_by_index[i] for i in section_map[field_name] if i in chunk_by_index
            ]
            if mapped_chunks:
                routes.append(
                    FieldRoute(
                        field_name=field_name,
                        field_spec=field_spec,
                        chunks=mapped_chunks,
                        source="section_map",
                    )
                )
                continue
            # Empty after filtering — fall through to heuristic

        # ── Heuristic path (original scorer) ──

        # look_in is a hard filter when any chunks match — schema authors
        # who declare a category are asserting the field lives there, so a
        # higher-scoring chunk in a different category should not win. Falls
        # back to the full pool when the category is absent from the doc,
        # so the field still gets routed somewhere.
        hints = field_spec.get("hints") or {}
        look_in = hints.get("look_in") or []
        candidate_chunks: list[Chunk] = chunks
        if look_in:
            matches = [c for c in chunks if c.category in look_in]
            if matches:
                candidate_chunks = matches

        # Score every candidate chunk for this field
        total_chunks = len(chunks)
        scored = []
        for chunk in candidate_chunks:
            score = _score_chunk(chunk, field_name, field_spec, total_chunks)
            if score > 0:
                scored.append((score, chunk))

        scored.sort(key=lambda x: x[0], reverse=True)
        top_chunks = [chunk for _, chunk in scored[:field_cap]]

        if top_chunks:
            source = "hint" if has_hints else "signal_inferred"
            routes.append(
                FieldRoute(
                    field_name=field_name,
                    field_spec=field_spec,
                    chunks=top_chunks,
                    source=source,
                )
            )
        else:
            # Nothing matched — broaden to any chunk with generic signals
            broadened = [c for c in chunks if c.signals]
            if broadened:
                routes.append(
                    FieldRoute(
                        field_name=field_name,
                        field_spec=field_spec,
                        chunks=broadened[:field_cap],
                        source="broadened",
                    )
                )
            else:
                # Last resort — first chunks
                routes.append(
                    FieldRoute(
                        field_name=field_name,
                        field_spec=field_spec,
                        chunks=chunks[:field_cap],
                        source="fallback",
                    )
                )

    return routes


def group_routes(routes: list[FieldRoute]) -> list[dict]:
    """Group fields that share the same chunks into extraction groups.
    Minimizes LLM calls — fields from the same chunk get extracted together."""
    groups: list[dict] = []
    used_fields: set[str] = set()

    for route in sorted(routes, key=lambda r: tuple(c.index for c in r.chunks)):
        if route.field_name in used_fields:
            continue

        chunk_indices = frozenset(c.index for c in route.chunks)

        # Find other fields that share the same chunks
        group_fields = [route]
        for other in routes:
            if other.field_name in used_fields or other.field_name == route.field_name:
                continue
            other_indices = frozenset(c.index for c in other.chunks)
            overlap = len(chunk_indices & other_indices) / max(len(chunk_indices), 1)
            if overlap >= 0.5:
                group_fields.append(other)

        # Collect unique chunks
        all_chunks: dict[int, Chunk] = {}
        for field_route in group_fields:
            for chunk in field_route.chunks:
                all_chunks[chunk.index] = chunk

        field_names = [f.field_name for f in group_fields]
        field_specs = {f.field_name: f.field_spec for f in group_fields}

        for name in field_names:
            used_fields.add(name)

        groups.append(
            {
                "fields": field_names,
                "field_specs": field_specs,
                "chunks": list(all_chunks.values()),
            }
        )

    return groups


def summarize_routing(routes: list[FieldRoute]) -> dict:
    """Summarize routing plan for logging."""
    plan = {}
    for route in routes:
        plan[route.field_name] = {
            "chunks": [f"{route.chunks[i].index}: {route.chunks[i].title}" for i in range(len(route.chunks))],
            "source": route.source,
        }
    return plan


# ── LLM Section Map ───────────────────────────────────────────────
#
# For long documents, ask the LLM which chunks contain which fields.
# This replaces the heuristic scorer with a content-aware assignment.


_CHUNK_PREVIEW_CHARS = 400


def _build_section_map_prompt(
    chunks: list[Chunk],
    schema_def: dict,
) -> str:
    """Build the prompt for the LLM section map pass.

    Sends a numbered list of chunk previews and a list of schema fields.
    Asks the LLM to return a JSON mapping of field_name → [chunk_indices].
    """
    fields = schema_def.get("fields", {})

    # Build chunk previews
    chunk_lines = []
    for chunk in chunks:
        preview = chunk.content[:_CHUNK_PREVIEW_CHARS].replace("\n", " ").strip()
        if len(chunk.content) > _CHUNK_PREVIEW_CHARS:
            preview += "..."
        chunk_lines.append(f"  [{chunk.index}] {chunk.title}: {preview}")
    chunk_list = "\n".join(chunk_lines)

    # Build field descriptions
    field_lines = []
    for name, spec in fields.items():
        ftype = spec.get("type", "string")
        desc = spec.get("description", "").strip().split("\n")[0]  # first line only
        field_lines.append(f"  - {name} ({ftype}): {desc}" if desc else f"  - {name} ({ftype})")
    field_list = "\n".join(field_lines)

    field_names = list(fields.keys())

    return f"""You are analyzing a document to determine which sections contain which fields.

## Document sections

{chunk_list}

## Fields to locate

{field_list}

## Instructions

For each field, identify which document sections (by index number) are most likely to contain that field's value. A field may appear in multiple sections, or not at all.

Return ONLY valid JSON — a single object mapping each field name to an array of section indices. If a field is not present in any section, map it to an empty array.

Example format:
{{
  "{field_names[0] if field_names else "field_name"}": [0, 1],
  "{field_names[1] if len(field_names) > 1 else "other_field"}": [3]
}}

JSON:"""


async def build_section_map(
    chunks: list[Chunk],
    schema_def: dict,
    provider: "ModelProvider",
) -> SectionMap | None:
    """Ask the LLM to map fields to chunks for a long document.

    Returns a dict of {field_name: [chunk_indices]} or None if the
    LLM call fails. The caller should fall back to heuristic routing
    when None is returned.
    """
    from .providers import ModelProvider  # avoid circular import at module level

    prompt = _build_section_map_prompt(chunks, schema_def)
    fields = schema_def.get("fields", {})

    try:
        raw = await provider.generate(prompt, json_mode=True)
        try:
            parsed = json.loads(raw)
        except json.JSONDecodeError:
            match = re.search(r"\{[\s\S]*\}", raw)
            if match:
                parsed = json.loads(match.group())
            else:
                print("[koji-extract] Section map: LLM returned invalid JSON")
                return None

        if not isinstance(parsed, dict):
            print("[koji-extract] Section map: LLM returned non-object")
            return None

        # Validate and normalize the response
        valid_indices = {c.index for c in chunks}
        section_map: SectionMap = {}
        for field_name in fields:
            indices = parsed.get(field_name, [])
            if not isinstance(indices, list):
                indices = [indices] if isinstance(indices, int) else []
            # Filter to valid chunk indices
            section_map[field_name] = [i for i in indices if isinstance(i, int) and i in valid_indices]

        return section_map

    except Exception as e:
        print(f"[koji-extract] Section map failed: {e}")
        return None


def needs_section_map(chunks: list[Chunk], max_chunks_per_field: int = 3) -> bool:
    """Determine whether a document is large enough to benefit from an LLM section map.

    The map pass adds one LLM call. It's worth it when the heuristic
    router would be forced to drop content — i.e., there are more
    chunks than any single field can see.
    """
    return len(chunks) > max_chunks_per_field * 2

File: api/src/extract/router.ts
/**
 * Phase 2: Field Router — match schema fields to the right chunks using schema hints.
 *
 * Faithful TypeScript port of services/extract/router.py. This module is
 * 100% deterministic (no LLM calls). It scores each field against each
 * chunk to find the most relevant chunks for extraction.
 */

import type { Chunk } from "./chunker";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface FieldRoute {
  fieldName: string;
  fieldSpec: Record<string, unknown>;
  chunks: Chunk[];
  source: "hint" | "signal_inferred" | "broadened" | "fallback" | "full_document" | "per_section";
}

export interface RouteGroup {
  fields: string[];
  fieldSpecs: Record<string, Record<string, unknown>>;
  chunks: Chunk[];
}

// ---------------------------------------------------------------------------
// Generic signal inference
// ---------------------------------------------------------------------------

/** When the schema has no hints, infer what to look for based on field type. */
const TYPE_SIGNAL_MAP: Record<string, string[]> = {
  date: ["has_dates"],
  number: ["has_dollar_amounts", "has_key_value_pairs"],
  string: ["has_key_value_pairs"],
  enum: ["has_key_value_pairs"],
  array: ["has_tables"],
};

// ---------------------------------------------------------------------------
// Chunk signal helper
// ---------------------------------------------------------------------------

/** Check whether a signal is present (truthy) on a chunk. */
function chunkHasSignal(chunk: Chunk, signal: string): boolean {
  return !!(chunk.signals as Record<string, unknown>)[signal];
}

/** Check whether a chunk has any truthy signal. */
function chunkHasAnySignal(chunk: Chunk): boolean {
  const signals = chunk.signals as Record<string, unknown>;
  return Object.values(signals).some((v) => !!v);
}

// ---------------------------------------------------------------------------
// Scoring
// ---------------------------------------------------------------------------

/**
 * Score how likely a chunk is to contain a field.
 * Uses schema hints or generic inference.
 *
 * Exported for testing — not part of the public API.
 */
export function scoreChunk(
  chunk: Chunk,
  fieldName: string,
  fieldSpec: Record<string, unknown>,
  totalChunks: number = 1,
): number {
  let score = 0;
  const hints = (fieldSpec.hints ?? {}) as Record<string, unknown>;

  // ── Hint-based scoring (user-defined, highest priority) ──

  const lookIn = (hints.look_in ?? []) as string[];
  if (lookIn.length > 0) {
    if (chunk.category && lookIn.includes(chunk.category)) {
      score += 15;
    }
  }

  const preferContains = (hints.prefer_contains ?? []) as string[];
  if (preferContains.length > 0) {
    const haystack = `${chunk.title} ${chunk.content}`.toLowerCase();
    for (const phrase of preferContains) {
      if (typeof phrase === "string" && phrase && haystack.includes(phrase.toLowerCase())) {
        score += 15;
        break;
      }
    }
  }

  const preferPosition = hints.prefer_position as string | undefined;
  if (preferPosition) {
    let frac: number;
    if (totalChunks <= 1) {
      frac = 0;
    } else {
      frac = chunk.index / (totalChunks - 1);
    }
    if (preferPosition === "top") {
      score += 10 * (1 - frac);
    } else if (preferPosition === "bottom") {
      score += 10 * frac;
    }
  }

  const patterns = (hints.patterns ?? []) as string[];
  if (patterns.length > 0) {
    const text = `${chunk.title} ${chunk.content.slice(0, 1500)}`.toLowerCase();
    for (const pattern of patterns) {
      try {
        if (new RegExp(pattern, "i").test(text)) {
          score += 8;
          break;
        }
      } catch {
        // Skip invalid patterns
      }
    }
  }

  const signalsHint = (hints.signals ?? []) as string[];
  if (signalsHint.length > 0) {
    for (const signal of signalsHint) {
      if (chunkHasSignal(chunk, signal)) {
        score += 4;
      }
    }
  }

  // If we had hints and scored, return early — hints are authoritative
  const hasHints = Object.keys(hints).length > 0;
  if (hasHints && score > 0) {
    return score;
  }

  // ── Generic inference (no hints provided) ──

  const fieldType = (fieldSpec.type as string) ?? "string";
  const inferredSignals = TYPE_SIGNAL_MAP[fieldType] ?? [];
  for (const signal of inferredSignals) {
    if (chunkHasSignal(chunk, signal)) {
      score += 2;
    }
  }

  // Field name appears in chunk title or content (fuzzy)
  const fieldWords = fieldName.replace(/_/g, " ").toLowerCase();
  if (chunk.title.toLowerCase().includes(fieldWords)) {
    score += 6;
  } else if (chunk.content.slice(0, 500).toLowerCase().includes(fieldWords)) {
    score += 3;
  }

  // Individual words from field name
  const words = fieldName.toLowerCase().split("_");
  const text = `${chunk.title} ${chunk.content.slice(0, 500)}`.toLowerCase();
  let wordHits = 0;
  for (const w of words) {
    if (w.length > 2 && text.includes(w)) {
      wordHits++;
    }
  }
  score += wordHits * 1.5;

  return score;
}

// ---------------------------------------------------------------------------
// Per-field max chunks
// ---------------------------------------------------------------------------

function fieldMaxChunks(fieldSpec: Record<string, unknown>, defaultMax: number): number {
  const hints = (fieldSpec.hints ?? {}) as Record<string, unknown>;
  const override = hints.max_chunks;
  if (typeof override === "number" && override > 0) {
    return override;
  }
  return defaultMax;
}

// ---------------------------------------------------------------------------
// Coverage-maximizing selection (per_section)
// ---------------------------------------------------------------------------

/** Default cap on distinct sections a `per_section` field will pull. */
const DEFAULT_MAX_SECTIONS = 24;

/**
 * The section a chunk belongs to, for coverage-maximizing selection. Generic:
 * the heading text (`title`) identifies a section — distinct parts of a large
 * package carry distinct headings. Falls back to `category`, then a shared
 * bucket, so untitled chunks don't each become their own "section".
 */
function sectionKey(chunk: Chunk): string {
  const title = (chunk.title ?? "").trim().toLowerCase();
  if (title) return `t:${title}`;
  if (chunk.category) return `c:${chunk.category}`;
  return "__untitled";
}

/** How much of a chunk's body to scan when anchor-matching a section. */
const ANCHOR_SCAN = 800;

/**
 * Compile a `section_anchor` hint (a regex string or a list of them) into
 * RegExps. Invalid patterns are skipped rather than failing the whole route.
 */
function compileSectionAnchors(raw: unknown): RegExp[] {
  const patterns = typeof raw === "string" ? [raw] : Array.isArray(raw) ? raw : [];
  const compiled: RegExp[] = [];
  for (const p of patterns) {
    if (typeof p !== "string" || !p) continue;
    try {
      compiled.push(new RegExp(p, "i"));
    } catch {
      // Skip a malformed pattern — a schema-author bug shouldn't drop the field.
    }
  }
  return compiled;
}

/**
 * Whether a chunk's section matches any anchor pattern. Matches against the
 * heading plus the head of the body so an anchor works whether the identifying
 * phrase is the title or the top line of the section.
 */
function matchesAnchor(chunk: Chunk, anchors: RegExp[]): boolean {
  const text = `${chunk.title ?? ""}\n${(chunk.content ?? "").slice(0, ANCHOR_SCAN)}`;
  return anchors.some((re) => re.test(text));
}

/**
 * Coverage-maximizing selection for an array field: instead of the globally
 * top-N chunks (which collapse onto the highest-scoring few and starve whole
 * sections on large multi-part documents), take the best-scoring chunk from
 * EACH distinct section so every qualifying section reaches the extractor.
 *
 * `scored` must be sorted by score descending — so the first chunk seen per
 * section is that section's best, and the Map preserves score-descending
 * section order for the safety cap. Returns chunks in document order (by
 * index) for a stable, readable prompt.
 */
/** Default cap on chunks pulled from a single section. Sections whose table
 *  spans multiple chunks need more than one, or rows in the dropped chunks are
 *  lost; the cap bounds context on pathologically large sections. */
const DEFAULT_CHUNKS_PER_SECTION = 4;

function selectCoverageMax(
  scored: Array<[number, Chunk]>,
  maxSections: number,
  chunksPerSection: number = DEFAULT_CHUNKS_PER_SECTION,
): { chunks: Chunk[]; sectionsFound: number } {
  // Group ALL chunks per section (score-descending within each), not just the
  // best one — a coverage-part/limits table that spans multiple chunks with the
  // same heading must keep every chunk, or the rows in the dropped chunks never
  // reach the model (the intra-section recall bug).
  const bySection = new Map<string, Chunk[]>();
  for (const [, chunk] of scored) {
    const key = sectionKey(chunk);
    const bucket = bySection.get(key);
    if (bucket) bucket.push(chunk);
    else bySection.set(key, [chunk]);
  }
  const sectionsFound = bySection.size;
  // Insertion order = score-descending by each section's best chunk, so slicing
  // keeps the highest-scoring sections when there are more than the cap.
  const chunks: Chunk[] = [];
  for (const bucket of [...bySection.values()].slice(0, maxSections)) {
    chunks.push(...bucket.slice(0, chunksPerSection));
  }
  chunks.sort((a, b) => a.index - b.index);
  return { chunks, sectionsFound };
}

/**
 * Expand a selected chunk set to include chunks within ±`radius` document
 * positions (by `index`), looked up in the FULL chunk set — so a neighbor in a
 * different category (one `look_in` filtered out) is still reachable. Lets a
 * field reach a value that sits in a chunk adjacent to the one holding its
 * label. Returns chunks in document order.
 */
function expandNeighbors(selected: Chunk[], all: Chunk[], radius: number): Chunk[] {
  const byIndex = new Map<number, Chunk>();
  for (const c of all) byIndex.set(c.index, c);
  const out = new Map<number, Chunk>();
  for (const c of selected) out.set(c.index, c);
  for (const c of selected) {
    for (let d = 1; d <= radius; d++) {
      for (const ni of [c.index - d, c.index + d]) {
        const n = byIndex.get(ni);
        if (n && !out.has(ni)) out.set(ni, n);
      }
    }
  }
  return [...out.values()].sort((a, b) => a.index - b.index);
}

// ---------------------------------------------------------------------------
// Route fields
// ---------------------------------------------------------------------------

/**
 * Route each schema field to the most relevant chunks.
 *
 * `maxChunksPerField` is the default cap. Individual fields can
 * override via `hints.max_chunks` in the schema.
 */
export function routeFields(
  schemaDef: Record<string, unknown>,
  chunks: Chunk[],
  maxChunksPerField: number = 3,
): FieldRoute[] {
  const fields = (schemaDef.fields ?? {}) as Record<string, Record<string, unknown>>;
  const routes: FieldRoute[] = [];

  for (const [fieldName, fieldSpec] of Object.entries(fields)) {
    const hasHints = !!fieldSpec.hints;
    const fieldCap = fieldMaxChunks(fieldSpec, maxChunksPerField);

    // look_in is a hard filter when any chunks match
    const hints = (fieldSpec.hints ?? {}) as Record<string, unknown>;
    const lookIn = (hints.look_in ?? []) as string[];
    let candidateChunks = chunks;
    if (lookIn.length > 0) {
      const matches = chunks.filter((c) => c.category && lookIn.includes(c.category));
      if (matches.length > 0) {
        candidateChunks = matches;
      }
    }

    // Score every candidate chunk for this field
    const totalChunks = chunks.length;
    const scored: Array<[number, Chunk]> = [];
    for (const chunk of candidateChunks) {
      const s = scoreChunk(chunk, fieldName, fieldSpec, totalChunks);
      if (s > 0) {
        scored.push([s, chunk]);
      }
    }

    scored.sort((a, b) => b[0] - a[0]);

    // `per_section`: coverage-maximizing selection. For an array field on a
    // large multi-part document, one chunk per distinct section reaches the
    // extractor — not just the globally top-N (which collapse onto the few
    // highest-scoring chunks and drop whole sections). Opt-in per field, so
    // small/monoline docs are unaffected (they simply have fewer sections).
    const perSection = hints.per_section === true;
    let topChunks: Chunk[];
    if (perSection && scored.length > 0) {
      const maxSections =
        typeof hints.max_sections === "number" && hints.max_sections > 0
          ? hints.max_sections
          : DEFAULT_MAX_SECTIONS;
      // `section_anchor` gates WHICH sections per_section visits: only sections
      // whose heading/top-of-body matches an anchor pattern get a representative
      // chunk. This stops per_section from over-producing spurious rows out of
      // boilerplate (product-catalog menus, "who is an insured" blocks) on large
      // packages. If the anchor matches nothing, fall back to all sections (a
      // too-narrow pattern shouldn't make a field silently vanish).
      const anchors = compileSectionAnchors(hints.section_anchor);
      let scopedScored = scored;
      if (anchors.length > 0) {
        const filtered = scored.filter(([, c]) => matchesAnchor(c, anchors));
        if (filtered.length > 0) {
          scopedScored = filtered;
        } else {
          console.warn(
            `[koji-extract] Route: field '${fieldName}' section_anchor matched no section — using all sections.`,
          );
        }
      }
      const chunksPerSection =
        typeof hints.chunks_per_section === "number" && hints.chunks_per_section > 0
          ? hints.chunks_per_section
          : DEFAULT_CHUNKS_PER_SECTION;
      const sel = selectCoverageMax(scopedScored, maxSections, chunksPerSection);
      topChunks = sel.chunks;
      if (sel.sectionsFound > maxSections) {
        console.warn(
          `[koji-extract] Route: field '${fieldName}' per_section capped ` +
            `${sel.sectionsFound} sections to ${maxSections} (raise hints.max_sections to include all).`,
        );
      }
    } else {
      topChunks = scored.slice(0, fieldCap).map(([, c]) => c);
    }

    // `neighbor_radius`: a value can land in a different chunk than its label
    // (e.g. a dense dec where "$1,691" and its "TOTAL DEPOSIT PREMIUM" label
    // split across a chunk boundary), so no single routed chunk holds both
    // anchor and value and the field is unreachable via look_in/prefer_contains
    // alone. When set, also pull chunks within ±radius document positions of
    // each selected chunk. Opt-in; pair with prefer_contains/patterns so the
    // label chunk is the one selected and its neighbor carries the value.
    const neighborRadius =
      typeof hints.neighbor_radius === "number" && hints.neighbor_radius > 0
        ? hints.neighbor_radius
        : 0;
    if (neighborRadius > 0 && topChunks.length > 0) {
      topChunks = expandNeighbors(topChunks, chunks, neighborRadius);
    }

    if (topChunks.length > 0) {
      const source = perSection ? "per_section" : hasHints ? "hint" : "signal_inferred";
      routes.push({
        fieldName,
        fieldSpec,
        chunks: topChunks,
        source,
      });
    } else {
      // Nothing matched — broaden to any chunk with generic signals
      const broadened = chunks.filter((c) => chunkHasAnySignal(c));
      if (broadened.length > 0) {
        routes.push({
          fieldName,
          fieldSpec,
          chunks: broadened.slice(0, fieldCap),
          source: "broadened",
        });
      } else {
        // Last resort — first chunks
        routes.push({
          fieldName,
          fieldSpec,
          chunks: chunks.slice(0, fieldCap),
          source: "fallback",
        });
      }
    }
  }

  return routes;
}

/**
 * Route every field to the full chunk set — used by adaptive routing when a
 * document is small enough that per-field selection only loses context. Each
 * field gets all chunks, so `groupRoutes` collapses them into a single group
 * (one full-document LLM call). Generic and document-type agnostic — the only
 * input is the chunk count, never the content or category.
 */
export function routeAllChunks(
  schemaDef: Record<string, unknown>,
  chunks: Chunk[],
): FieldRoute[] {
  const fields = (schemaDef.fields ?? {}) as Record<string, Record<string, unknown>>;
  return Object.entries(fields).map(([fieldName, fieldSpec]) => ({
    fieldName,
    fieldSpec,
    chunks,
    source: "full_document" as const,
  }));
}

// ---------------------------------------------------------------------------
// Group routes
// ---------------------------------------------------------------------------

/**
 * Group fields that share the same chunks into extraction groups.
 * Minimizes LLM calls — fields from the same chunk get extracted together.
 */
export function groupRoutes(routes: FieldRoute[]): RouteGroup[] {
  const groups: RouteGroup[] = [];

  // Some fields extract in isolation — each becomes its own singleton group and
  // takes no part in the overlap-based grouping below (as either a seed or a
  // candidate). Two cases:
  //
  //  • per_section fields, whose chunk set is deliberately expanded to one chunk
  //    per section (potentially spanning the whole document). Folding that union
  //    into a shared group would make every sibling extract against all those
  //    sections — a scalar sharing one section chunk would inherit the rest and
  //    mis-extract (e.g. a carrier name stamped on each coverage-part header).
  //
  //  • fields with `hints.isolate: true`. A grouped field is extracted in one
  //    LLM call over the UNION of the group's chunks, alongside every other
  //    member's instructions — so its output depends on siblings' routing and
  //    flips when an unrelated field's hint changes. `isolate` pins a critical
  //    field to exactly its own routed chunks and its own instruction, making it
  //    stable with respect to the rest of the schema.
  const isIsolated = (r: FieldRoute): boolean =>
    r.source === "per_section" ||
    ((r.fieldSpec.hints as Record<string, unknown> | undefined)?.isolate === true);

  const groupableRoutes = routes.filter((r) => !isIsolated(r));
  for (const route of routes) {
    if (!isIsolated(route)) continue;
    groups.push({
      fields: [route.fieldName],
      fieldSpecs: { [route.fieldName]: route.fieldSpec },
      chunks: [...route.chunks],
    });
  }

  const usedFields = new Set<string>();

  const sorted = [...groupableRoutes].sort((a, b) => {
    const aIndices = a.chunks.map((c) => c.index);
    const bIndices = b.chunks.map((c) => c.index);
    for (let i = 0; i < Math.max(aIndices.length, bIndices.length); i++) {
      const ai = aIndices[i] ?? -1;
      const bi = bIndices[i] ?? -1;
      if (ai !== bi) return ai - bi;
    }
    return 0;
  });

  for (const route of sorted) {
    if (usedFields.has(route.fieldName)) continue;

    const chunkIndices = new Set(route.chunks.map((c) => c.index));

    // Find other fields that share the same chunks
    const groupFields: FieldRoute[] = [route];
    for (const other of groupableRoutes) {
      if (usedFields.has(other.fieldName) || other.fieldName === route.fieldName) {
        continue;
      }
      const otherIndices = new Set(other.chunks.map((c) => c.index));
      let overlapCount = 0;
      for (const idx of chunkIndices) {
        if (otherIndices.has(idx)) overlapCount++;
      }
      const overlap = overlapCount / Math.max(chunkIndices.size, 1);
      if (overlap >= 0.5) {
        groupFields.push(other);
      }
    }

    // Collect unique chunks
    const allChunks = new Map<number, Chunk>();
    for (const fieldRoute of groupFields) {
      for (const chunk of fieldRoute.chunks) {
        allChunks.set(chunk.index, chunk);
      }
    }

    const fieldNames = groupFields.map((f) => f.fieldName);
    const fieldSpecs: Record<string, Record<string, unknown>> = {};
    for (const f of groupFields) {
      fieldSpecs[f.fieldName] = f.fieldSpec;
    }

    for (const name of fieldNames) {
      usedFields.add(name);
    }

    groups.push({
      fields: fieldNames,
      fieldSpecs,
      chunks: [...allChunks.values()],
    });
  }

  return groups;
}

// ---------------------------------------------------------------------------
// Summarize routing (for logging)
// ---------------------------------------------------------------------------

export function summarizeRouting(routes: FieldRoute[]): Record<string, unknown> {
  const plan: Record<string, unknown> = {};
  for (const route of routes) {
    plan[route.fieldName] = {
      chunks: route.chunks.map((c) => `${c.index}: ${c.title}`),
      source: route.source,
    };
  }
  return plan;
}
